bubblewrap.c

```diff
@@ -1096,7 +1100,21 @@ privileged_op (int         privileged_op_socket,
 
     case PRIV_SEP_OP_TMPFS_MOUNT:
       {
-        cleanup_free char *mode = xasprintf ("mode=%#o", perms);
+        cleanup_free char *mode;
+
+        /* If we are given a size for the tmpfs, validate that it doesn't
+         * contain any special characters. */
+        const size_t arg2_len = strlen (arg2);
+        if (arg2_len > 0) {
+          for (size_t i = 0; i < arg2_len; i++) {
+            if (!isalnum (arg2[i]))
```

Comment:
Rather than accepting arbitrary letters and digits here, I wonder whether it'd be better to require the caller to specify the size as a simple integer (in bytes), and move any additional syntactic sugar into a parent process that is never setuid or privileged? That's in line with what we do for file modes, for example.

Reply:
This makes sense. However, the advantage of the current setup is that the parsing of the size specification is done by the tmpfs implementation, saving code here in bwrap.

Are you suggesting we do the parsing of the size spec ourselves, when reading the options?

Reply:
The reason I'm hesitant here is that we're handling strings (red flag) in C (red flag) in code that might be running setuid root (biggest red flag) and potentially passing them straight to the kernel (red flag) - so I want to minimize the amount of string handling in our own code, but also not let people pass unexpected strings to the kernel, where they will be applied to a tmpfs with full privileges.

When bwrap isn't setuid, we're not actually doing anything that the kernel wouldn't allow normally, but when bwrap *is* setuid, the sysadmin is trusting us to not pass anything "bad" to the kernel. That makes me cautious about this PR, which would accept `--size r2d2` and pass it directly to the kernel.

However, the more parsing and validation we do, the more string manipulation it implies, while potentially setuid root, in a language that is notoriously not memory-safe and has notoriously bad string APIs. A lot of oddities of bwrap's command-line, like `--perms` being a separate option, `--foo=bar` not being accepted, and `--bind` not having a `--bind=source:dest:flags` syntax like Docker's `-v`, come from the desire to minimize scary string manipulation in bwrap.

We can't really even mitigate C's lack of safe string handling by linking extra libraries, because if we do, we're trusting the entire library to be setuid-safe, and in practice even the most carefully-written libraries usually aren't.

So, what I initially had in mind was to only accept a bare number in bytes, read using `strtoul()`: something like `--size 12345`, while rejecting `--size 12345M` (which your PR would accept), `--size 12%`, or `--size r2d2` (which your PR would accept but probably shouldn't). If there's any clever human-friendly parsing happening, my idea was that something "larger" than bwrap - like Flatpak, or libgnome-desktop, or your language interpreter environment that is using bwrap internally - should be doing that clever parsing, and then passing the result in to the lower-level bwrap tool as a byte-count.

It would be easy to scope-creep from a plain unsigned integer to having an unsigned integer followed by up to one of `m`, `k`, `g` or `%`, which I think is basically what the kernel accepts, and it wouldn't even be *that* much code; but at what point are we doing enough that there's a danger of some subtle root-privilege-escalation vulnerability?

Reply:
Perhaps the answer is to only accept `--size` when unprivileged, and then do fairly minimal validation.

Reply:
> getting the amount of available RAM is slightly more involved. My free seems to get it from /proc/meminfo, which looks parseable, but I don't know if going there is a good idea for bubblewrap

That's the opposite of what I want: bwrap definitely shouldn't be rummaging in `/proc`. I'm trying to minimize bwrap's attack surface, and making it responsible for additional syntactic sugar would work against that.

Reply:
Thanks for the thorough analysis; I agree with you. My experience with writing security-sensitive code is limited (ahem), which is why the red flags did not really trigger for me as much as they maybe should. Code review is good.

> Perhaps the answer is to only accept `--size` when unprivileged, and then do fairly minimal validation.

For my use-case, only accepting `--size` when unprivileged would be fine, and only accepting a simple byte count would be fine too. And that scenario is basically the safest of the alternatives you describe. However, I would consider `strtoull()` to be fine in setuid code, too; after all, bwrap already uses `strtoul()` for `--perms` currently.

Do I change to code to accept a simple number of bytes in unprivileged mode only, or also under setuid?